File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
def solution(string):
    top = -1        # 열린 괄호 갯수 확인
    
    for s in string:
        if s == '(':
            top += 1
        
        else:       # 닫힌 괄호일 때
            if top > -1:        # 열린 괄호가 있으면
                top -= 1        # 열린 괄호 갯수 -1
            else:
                return False  # 열린 괄호가 없으면 False
                
    if top > -1:        # 남은 괄호가 있으면 false
        return False
    else:
        return True

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
class Solution {
    boolean solution(String s) {
        boolean answer = true;
        int top = -1;
        
        for (int i = 0; i < s.length(); i++) {
            char bracket = s.charAt(i);
            if (bracket == '(') {
                top++;
            } else {
                if (top < 0) {      // 닫는 괄호 나왔으나 여는 괄호 없는 경우
                    answer = false;
                    break;
                } else {
                    top--;
                }
            }
        }
        
        if (top > -1) {         // 문자열 끝났을 때 여는 괄호 남아있는 경우
            answer = false;
        }

        return answer;
    }
}